From 5ed9dec4c06954c39f9a4923366170fad056ccf3 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:07:28 -0400
Subject: [usb] uas: fix error handling during scsi_scan()

Message-id: <1410887463-20674-377-git-send-email-dzickus@redhat.com>
Patchwork-id: 95447
O-Subject: [RHEL7 PATCH 376/591] uas: fix error handling during scsi_scan()
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit c637f1fa7b0452b71eebd35d00906d371c04714e
Author: Oliver Neukum <oneukum@suse.de>
Date:   Fri Mar 28 11:29:25 2014 +0100

    uas: fix error handling during scsi_scan()

    intfdata is set only after scsi_scan(). uas_pre_reset() however
    needs intfdata to be valid and will follow the NULL pointer
    killing khubd. intfdata must be preemptively set before the
    host is registered and undone in the error case.

    Signed-off-by: Oliver Neukum <oliver@neukum.org>
    Reviewed-by: Hans de Goede <hdegoede@redhat.com>
    Acked-by: Hans de Goede <hdegoede@redhat.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/storage/uas.c b/drivers/usb/storage/uas.c
index 8f42226..fcab9b7 100644
--- a/drivers/usb/storage/uas.c
+++ b/drivers/usb/storage/uas.c
@@ -1096,16 +1096,17 @@ static int uas_probe(struct usb_interface *intf, const struct usb_device_id *id)
  if (result)
   goto free_streams;
 
+ usb_set_intfdata(intf, shost);
  result = scsi_add_host(shost, &intf->dev);
  if (result)
   goto free_streams;
 
  scsi_scan_host(shost);
- usb_set_intfdata(intf, shost);
  return result;
 
 free_streams:
  uas_free_streams(devinfo);
+ usb_set_intfdata(intf, NULL);
 set_alt0:
  usb_set_interface(udev, intf->altsetting[0].desc.bInterfaceNumber, 0);
  if (shost)
-- 
1.7.1